### PR TITLE
Reimplement the PluginGenerator based on dir templating

### PR DESCRIPTION
**Pull request by mfurtak**
Moves all of the plugin generator template files from a flat directory into one that is actually shaped like the intended destination structure.

This allows the PluginGenerator to stop encoding knowledge about the intended structure of the plugin, and just focus on copying and template logic.

The template directory uses %-wrapped variables to allow directory and file names to have values inserted at run time. Those variable names are used as accessor calls into the `PluginInfo` object to get replacement values.

The tests have been changed to use a plugin name containing an underscore and have been corrected to check for expected values based on `fastlane_class` transform of the plugin name

### CODE REVIEW

**Comment by mfurtak** [comment]
@KrauseFx @ohwutup @mpirri

File: fastlane/lib/fastlane/plugins/plugin_generator.rb [hunk 1]
```diff
@@ -1,33 +1,32 @@
...
-
 module Fastlane
+  # Generates a sample plugin by traversing a template directory structure
+  # and reproducing it in a destination location. At the same time, it runs
+  # variable replacements on directory names, file names, and runs ERB
+  # templating on file contents whose names end with '.erb'.
+  #
+  # Directory and file name variable replacements are defined like: %gem_name%
+  # The text between the percent signs will be used to invoke an accessor
+  # method on the PluginInfo object to get the replacement value.
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/plugins/plugin_generator.rb]: ❤️ comments

File: fastlane/spec/plugins_specs/plugin_generator_spec.rb [hunk 2]
```diff
@@ -8,7 +8,7 @@
 
 describe Fastlane::PluginGenerator do
   describe '#generate' do
-    let(:plugin_info) { Fastlane::PluginInfo.new('tester', 'Fabricio Devtoolio', 'fabric.devtools@gmail.com', 'summary') }
+    let(:plugin_info) { Fastlane::PluginInfo.new('tester_thing', 'Fabricio Devtoolio', 'fabric.devtools@gmail.com', 'summary') }
```
Comment (KrauseFx) [inline fastlane/spec/plugins_specs/plugin_generator_spec.rb]: I love how the tests have hardly changed, they're all passing and the implementation was completely refactored! Great job 👍

**Comment by KrauseFx** [comment]
This looks great, I love this change, it will enable us to move much faster in the future. This way we can also easily copy over the fastlane badge, right? Should it be moved to the `template` directory?

File: fastlane/spec/plugins_specs/plugin_generator_spec.rb (hunk 2 above)
Comment (mfurtak) [inline fastlane/spec/plugins_specs/plugin_generator_spec.rb]: Yes, thanks! Having robust tests made this super easy to do with confidence 🙌

**Comment by mfurtak** [comment]
> This way we can also easily copy over the fastlane badge, right? Should it be moved to the template directory?

There was discussion in the past about whether the badge should be a part of each plugin project, or whether it should be "hot-linked" from the main fastlane repo. My opinion was that it would make it easier to improve/change the badge in the future if all plugin projects just referred to the badge via URL.

In that case maybe it makes sense to move the badge to the main fastlane assets folder, and update the link in the plugin readme?

**Comment by KrauseFx** [comment]
Ahh, right! I'm okay with having it in a central place instead 👍 Do we link to the image in the README already?

**Comment by mfurtak** [comment]
Yep, the plugin's README already contains a reference, but it will need to be updated. I'll make both of those changes in a new PR.

**Comment by KrauseFx** [comment]
🚀

**Pull request merged**